evidence/reactor.go

```diff
@@ -21,233 +46,346 @@ const (
 	// Most evidence should be committed in the very next block that is why we wait
 	// just over the block production rate before sending evidence again.
 	broadcastEvidenceIntervalS = 10
-	// If a message fails wait this much before sending it again
-	peerRetryMessageIntervalMS = 100
 )
 
 // Reactor handles evpool evidence broadcasting amongst peers.
 type Reactor struct {
-	p2p.BaseReactor
-	evpool   *Pool
-	eventBus *types.EventBus
+	service.BaseService
+
+	evpool       *Pool
+	eventBus     *types.EventBus
+	evidenceCh   *p2p.Channel
+	peerUpdates  *p2p.PeerUpdatesCh
+	closeCh      chan struct{}
+	peerRoutines map[string]chan struct{}
+
+	mtx    tmsync.RWMutex
+	peerWG sync.WaitGroup
 }
 
-// NewReactor returns a new Reactor with the given config and evpool.
-func NewReactor(evpool *Pool) *Reactor {
-	evR := &Reactor{
-		evpool: evpool,
+// NewReactor returns a reference to a new evidence reactor, which implements the
+// service.Service interface. It accepts a p2p Channel dedicated for handling
+// envelopes with EvidenceList messages.
+func NewReactor(
+	logger log.Logger,
+	evidenceCh *p2p.Channel,
+	peerUpdates *p2p.PeerUpdatesCh,
+	evpool *Pool,
+) *Reactor {
+	r := &Reactor{
+		evpool:       evpool,
+		evidenceCh:   evidenceCh,
+		peerUpdates:  peerUpdates,
+		closeCh:      make(chan struct{}),
+		peerRoutines: make(map[string]chan struct{}),
 	}
-	evR.BaseReactor = *p2p.NewBaseReactor("Evidence", evR)
-	return evR
+
+	evpool.SetLogger(logger)
```

Comment:
Setup of injected dependencies should be done by the caller, e.g. it may already have set its own preferred logger for the evidence pool and we shouldn't override that with the reactor logger.

Reply:
I would rather change the pool's constructor API then -- this is a codesmell IMHO.